src/client/actions/ThreadListSync.js

```diff
@@ -0,0 +1,68 @@
+'use strict';
+
+const Action = require('./Action');
+const Collection = require('../../util/Collection');
+const { Events } = require('../../util/Constants');
+
+class ThreadListSyncAction extends Action {
+  handle(data) {
+    const client = this.client;
+
+    const guild = client.guilds.cache.get(data.guild_id);
+    if (!guild) return {};
+
+    if (!data.channels_ids) {
+      guild.channels.cache.forEach(channel => {
+        this.removeStale(channel);
+      });
+    } else {
+      data.channel_ids.forEach(id => {
+        const channel = client.channels.resolve(id);
+        if (channel) this.removeStale(channel);
+      });
+    }
+
+    Object.values(data.members).forEach(rawMember => {
+      // Discord sends the thread id as id in this object
+      const thread = client.channels.cache.get(rawMember.id);
+      if (thread) {
+        thread.members.add(rawMember);
+      }
+    });
+
+    const syncedThreads = new Collection();
+    data.threads.forEach(rawThread => {
+      const thread = client.channels.add(rawThread);
+      syncedThreads.set(thread.id, thread);
+    });
+
+    data.members.forEach(rawMember => {
+      // Discord sends the thread id as id in this object
+      const thread = client.channels.cache.get(rawMember.id);
+      if (thread) {
+        thread.members.add(rawMember);
+      }
+    });
```

Comment:
We iterate over the thread members TWICE, one assumes it can be an array or an object, the other that it's definitely an array. Can we do one iteration after you "sync threads"? Would probably achieve the exact same thing

Reply:
Looks like this accidentally sneaked in at this commit: https://github.com/discordjs/discord.js/pull/5570/commits/1d7455b9abffaf78d0be6d87285395183b42e4ed

Reply:
ahh, messed up the rebase a bit I suppose